### PR TITLE
[CN-Test-Gen] Purely unfolding implementation

### PR DESCRIPTION
**Pull request by ZippeyKeys12**
Add an implementation of test generation.

This version generates Google Test harnesses with RapidCheck generators, one per path through the precondition.
It unfolds recursive predicates up to the specified max depth.

I still have to add some tests for CI.

### CODE REVIEW

File: backend/cn/main.ml [hunk 1]
```diff
@@ -433,6 +445,20 @@ let lemmata =
 
 end
 
+module TestGenerationFlags = struct
```
Comment (dc-mak) [inline backend/cn/main.ml]: Convention is Snake_case.

File: backend/cn/main.ml [hunk 2]
```diff
@@ -433,6 +445,20 @@ let lemmata =
...
+let output_test_dir =
+  let doc = "[Experimental] Generate test suites and place them in the provided directory" in
+  Arg.(value & opt (some string) None & info ["output-test-dir"] ~docv:"FILE" ~doc)
+
```
Comment (dc-mak) [inline backend/cn/main.ml:452]: In a **follow-up PR**, can you turn this into a subcommand `cn generate-tests`?

File: backend/cn/testGeneration.ml [hunk 3]
```diff
@@ -0,0 +1,957 @@
+module AT = ArgumentTypes
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml:7]: Can you please provide a two line summary of this file at the top here?

File: backend/cn/testGeneration.ml [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+        , IT.fresh_named
+            (bt_of_ctype (Cerb_location.other __FUNCTION__) ty)
+            (Pp_symbol.to_string_pretty sym)
+            (Cerb_location.other __FUNCTION__)
+          |> snd ) )
+      :: vars
+    , ms )
+;;
+
+let ( >>= ) (x : 'a list) (f : 'a -> 'b list) : 'b list = List.flatten (List.map f x)
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: OCaml convention for monads is `let (let*) x f = ...`.

File: backend/cn/testGeneration.mli [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module CF = Cerb_frontend
+
+type test_framework = GTest
+
+val main
+  :  string
+  -> string
+  -> int
```
Comment (dc-mak) [inline backend/cn/testGeneration.mli]: Can you please use named parameters for the two strings (and maybe the int)?
Can you also add a doc comment for this?

File: backend/cn/testGeneration.ml [hunk 6]
```diff
@@ -0,0 +1,957 @@
...
+         (fun (e, x) ->
+           "(*"
+           ^ Pp_utils.to_plain_pretty_string (IT.pp e)
+           ^ ") -> "
+           ^ Pp_symbol.to_string_pretty x)
+         locs)
+  ^ " }"
+;;
+
+type members = (Symbol.sym * (string * (Ctype.ctype * Symbol.sym)) list) list
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: Using a record and some field names would make this more readable.

File: backend/cn/testGeneration.ml [hunk 7]
```diff
@@ -0,0 +1,957 @@
...
+  =
+  match ty with
+  | Ctype (_, Struct n) ->
+    (match List.assoc weak_sym_equal n sigma.tag_definitions with
+     | _, _, StructDef (membs, _) ->
+       let f (Symbol.Identifier (_, id), (_, _, _, ty)) =
+         let sym' = Symbol.fresh () in
+         ( ( sym'
+           , ( ty
+             , IT.IT
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: Please use smart constructors such as `IT.sym_`.

File: backend/cn/testGeneration.ml [hunk 8]
```diff
@@ -0,0 +1,957 @@
...
+module IT = IndexTerms
+module LS = LogicalSorts
+module RET = ResourceTypes
+module LC = LogicalConstraints
+module LAT = LogicalArgumentTypes
+module CF = Cerb_frontend
+open CF
+
+let weak_sym_equal s1 s2 =
+  String.equal (Pp_symbol.to_string_pretty s1) (Pp_symbol.to_string_pretty s2)
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: Can you add a comment explaining why you need equality on `to_string_pretty`?

File: backend/cn/main.ml [hunk 9]
```diff
@@ -217,6 +228,7 @@ let main
       print_log_file ("mucore", MUCORE prog5);
       let paused = Typing.run_to_pause Context.empty (Check.check_decls_lemmata_fun_specs prog5) in
       Result.iter_error handle_error (Typing.pause_to_result paused);
+      maybe_generate_tests output_test_dir filename test_max_depth testing_framework ail_prog prog5;
       begin match output_decorated with
       | None ->
         Typing.run_from_pause (fun paused -> Check.check paused lemmata) paused
```
Comment (dc-mak) [inline backend/cn/main.ml]: This will run the proof checking after generating tests, which is likely not what we want.

**Review by dc-mak — CHANGES_REQUESTED** [review]
Looking reasonable, some minor suggestions. I know some steps for the CI are a pain to add but a few things in CI will make everyone's development process smoother.

File: backend/cn/testGeneration.ml (hunk 4 above)
Comment (ZippeyKeys12) [reply backend/cn/testGeneration.ml]: The rest of CN uses `let@`, so I changed it to that instead.

File: backend/cn/main.ml (hunk 9 above)
Comment (ZippeyKeys12) [reply backend/cn/main.ml]: I figured that could be dealt with after we split things into subcommands...

File: backend/cn/main.ml (hunk 2 above)
Comment (ZippeyKeys12) [reply backend/cn/main.ml:452]: That's the plan 👍

File: backend/cn/main.ml (hunk 9 above)
Comment (ZippeyKeys12) [reply backend/cn/main.ml]: Though I suppose for CI tests, why waste the time.

File: backend/cn/testGeneration.ml [hunk 10]
```diff
@@ -1,3 +1,9 @@
+(** Test Generation
+    Handles all parts of test generation from
+    gathering constraints to building
+    generators and finally exporting C++ code.
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: This comment doesn't really explain what "test generation" means, or what the high-level purpose of it being in CN and using the specifications/types.

File: backend/cn/testGeneration.ml [hunk 11]
```diff
@@ -46,35 +61,33 @@ let string_of_locations (locs : locations) : string =
...
   ^ " }"
 ;;
 
-type members = (Symbol.sym * (string * (Ctype.ctype * Symbol.sym)) list) list
+(** Tracks indirection for a struct's member [name],
+    where [car] carries its value of type [cty].
+    **)
+type member =
+  { name : string (** The name of the member *)
+  ; car : Sym.sym (** The name of the carrier*)
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: Please use the full word `carrier` (or something more descriptive).

**Review by dc-mak — APPROVED** [review]
I'm approving, though I would like to see the changes requested in the follow-up PR (CI, subcommands, rename `car` to `carrier`, and a more explanatory file explainer).

File: backend/cn/testGeneration.ml [hunk 12]
```diff
@@ -730,54 +730,59 @@ let compile ((vars, ms, locs, cs) : goal) : gen_context =
...
+       | Binop (LE, e1, e2) | Binop (LEPointer, e1, e2) ->
+         "(" ^ codify_it e1 ^ " <= " ^ codify_it e2 ^ ")"
+       | Binop (EQ, e1, e2) -> "(" ^ codify_it e1 ^ " == " ^ codify_it e2 ^ ")"
+       | Binop (Mod, e1, e2) -> "(" ^ codify_it e1 ^ " % " ^ codify_it e2 ^ ")"
+       | ITE (e1, e2, e3) ->
+         "(" ^ codify_it e1 ^ " ? " ^ codify_it e2 ^ " : " ^ codify_it e3 ^ ")"
+       (* *)
+       | _ -> failwith __FUNCTION__)
+  with
+  | Failure str when String.equal __FUNCTION__ str -> None
```
Comment (dc-mak) [inline backend/cn/testGeneration.ml]: Catching `Failure` with string equality is brittle. I think the best way to proceed would be to wrap this in the option monad (you can do `let (let<punctuation>)` locally for just this function. Alternatively, you can declare a local exception `let exception No_codify_it` and `raise No_codify_it` instead.

**Review by dc-mak — CHANGES_REQUESTED** [review]

File: backend/cn/testGeneration.ml (hunk 12 above)
Comment (ZippeyKeys12) [reply backend/cn/testGeneration.ml]: I didn't know about local exceptions, I've switched to one

**Review by dc-mak — APPROVED** [review]

**Comment by dc-mak** [comment]
Ah damnit the squashed commit message didn't include "This version generates Google Test harnesses with RapidCheck generators, one per path through the precondition. It unfolds recursive predicates up to the specified max depth." which would have been a better summary than what I wrote. Oh well. Next time I'll just ask you to squash it instead.

**Pull request merged**